src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -259,5 +291,54 @@ async def generate_all_tool_lockfiles(
     return ToolLockGoal(exit_code=0)
 
 
+BEGIN_LOCKFILE_HEADER = b"# --- BEGIN LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---"
+END_LOCKFILE_HEADER = b"# --- END LOCKFILE METADATA ---"
+
+
+# Lockfile metadata for headers
+def validated_lockfile_content(req: PythonLockfileRequest, content: bytes) -> bytes:
+    return b"%b\n%b" % (lockfile_metadata(req.hex_digest), content)
+
+
+def lockfile_metadata(invalidation_digest: str) -> bytes:
+    """Produces a metadata bytes object for including at the top of a lockfile.
+
+    Currently, this only consists of an invalidation digest for the file, which is used when Pants
+    consumes the lockfile during builds.
+    """
+    return (
+        b"""
+%(BEGIN_LOCKFILE_HEADER)b
+# invalidation digest: %(invalidation_digest)s
+%(END_LOCKFILE_HEADER)b
+    """
+        % {
+            b"BEGIN_LOCKFILE_HEADER": BEGIN_LOCKFILE_HEADER,
+            b"invalidation_digest": invalidation_digest.encode("ascii"),
+            b"END_LOCKFILE_HEADER": END_LOCKFILE_HEADER,
+        }
+    ).strip()
+
+
+def read_lockfile_metadata(contents: bytes) -> dict[str, str]:
```

Comment:
This is a good starting point. In the future, as we add more metadata, it would be great to use either `TypedDict` or convert into a dataclass so that we all callers of this function have a more clear API. https://www.python.org/dev/peps/pep-0589/